File: youtube_summarizer/schemas.py
from pydantic import BaseModel, Field, field_validator

from .utils import s2hk


class Chapter(BaseModel):
    """Represents a single chapter in the video summary."""

    title: str = Field(
        description="A concise chapter heading.",
    )
    description: str = Field(
        description="A substantive chapter description grounded in the content. Include key facts (numbers/names/steps) when present. Avoid meta-language like 'the video', 'the author', 'the speaker says'—state the content directly.",
    )
    start_time: str | None = Field(
        None,
        description="Optional chapter start timestamp in the format MM:SS.",
    )
    end_time: str | None = Field(
        None,
        description="Optional chapter end timestamp matching the same format as start_time.",
    )

    @field_validator("title", "description")
    def convert_string_to_hk(cls, value: str) -> str:
        """Convert string fields to Traditional Chinese if needed (helper hook)."""
        return s2hk(value)


class Summary(BaseModel):
    """Complete analysis of a YouTube video."""

    overview: str = Field(
        description="An end-to-end summary of the whole content (main thesis + arc), written in direct statements without meta-language.",
    )
    chapters: list[Chapter] = Field(
        min_length=1,
        description="Chronological, non-overlapping chapters covering the core content.",
    )

    @field_validator("overview")
    def convert_string_to_hk(cls, value: str) -> str:
        """Convert string fields to Traditional Chinese."""
        return s2hk(value)

    def to_text(self) -> str:
        """Format the summary as a readable string."""
        lines = [
            "=" * 80,
            "SUMMARY:",
            "=" * 80,
            f"\nOverview:\n{self.overview}",
            f"\nChapters ({len(self.chapters)}):",
        ]

        for i, chapter in enumerate(self.chapters, 1):
            lines.append(f"\n  Chapter {i}: {chapter.title}")
            lines.append(f"    Summary: {chapter.description}")
            if chapter.start_time or chapter.end_time:
                time_range = f"{chapter.start_time or '?'} - {chapter.end_time or '?'}"
                lines.append(f"    Time: {time_range}")

        return "\n".join(lines)

File: youtube-summarizer/schemas.ts
import { z } from "zod";

import { s2hk } from "./utils.js";

export const ChapterSchema = z.object({
  title: z.string().transform((value) => s2hk(value)),
  description: z.string().transform((value) => s2hk(value)),
  start_time: z.string().nullable().optional(),
  end_time: z.string().nullable().optional(),
});

export const SummarySchema = z.object({
  overview: z.string().transform((value) => s2hk(value)),
  chapters: z.array(ChapterSchema).min(1),
});

export type Chapter = z.output<typeof ChapterSchema>;
export type Summary = z.output<typeof SummarySchema>;

export function parseSummaryJson(raw: string): Summary {
  const parsed = JSON.parse(raw) as unknown;
  return SummarySchema.parse(parsed);
}

export function parseSummary(input: unknown): Summary {
  return SummarySchema.parse(input);
}
